README
TABLE OF CONTENTS
=================

        1. About Canute
        2. Usage
        3. Compilation
                3.1 Hasefroch
        4. Protocol enhancements
                4.1 File modification time
                4.2 Executable bit
        5. Source code files
        6. Credits



1. ABOUT CANUTE
===============

Canute is a small command line utility to transfer files and directories
over the network.  It does not have authentication nor any other kind of
privileges.  It can be handy in a number of situations, specially on LAN.

For example, imagine you are with a friend and he has a big CD/DVD image
you want (whatever Linux distro).  Depending on the operating systems you
have, there are some ways of doing the copy:

   - FTP: Set up a server (you need root privileges for that) and then
     allow anonymous upload, create a new account for your friend or give
     him access using your account.  Don't forget a good FTP client

   - HTTP: Maybe handier than FTP but upload is a bit trickier.

   - SFTP/SSH: This is account based so give new accounts or your own.

   - RSync: If you want to use it two way you need to transfer over SSH,
     then you are in the same situation as before.  Otherwise both of you
     should configure a server with the appropriate modules (which, again,
     requires root privileges).

   - NFS: Hasefroch does not easily support it.  This is not as hard as it
     used to be thanks to the modern Linux distributions.  But you need at
     least two servers: the port mapper and the nfsd.  Not mentioning the
     kernel support, the permission grant from the server
     (/etc/hosts.allow), etc...   And it is quite slow for big files.

   - SMB: Very similar to NFS but slower.

   - Finally, you could also waste a CD/DVD burning that ISO image and
     giving it to your friend.

With Canute you use the same binary for all tasks: sending or receiving.
No need for installation, just the damn binary.



2. USAGE
========

There are two pairs of concepts: server-client and sender-receiver.  The
server-client couple only defines who (of the two peers) has to execute the
application first.  The sender-receiver defines the direction of the
transfer.

Let's have host A and host B.  The following scenarios show the commands
and the order in which must be executed.

==> Scenario 1.  A sends files to B.  This is the usual way of working.

host_A$ canute send file1 file2 ...
host_B$ canute get host_A

==> Scenario 2.  A sends files to B, but A is behind a firewall and cannot
    open ports (NAT/PAT)

host_B$ canute getserv
host_A$ canute sendto host_B file1 file2 ...

We see how server-client roles can be switched independently from
sender-receiver roles.  The server is the peer who waits for the client to
be connected.  The server is not persistent and the port is closed when it
finishes.  That means you do not have to worry about leaving that port
opened on your router if you want to do Internet transfers.

You can also choose a different port than the default (1121/tcp) specifying
it after the Canute sub-command, like this

host_A$ canute send:5030 file1 file2 ...
host_B$ canute get:5030 host_A

Or else:

host_B$ canute getserv:5030
host_A$ canute sendto:5030 host_B file1 file2 ...

Note that if a directory is found, in the command line, then is sent
recursively.



3. COMPILATION
==============

Canute can now compile in many UNIX flavours, as well as in Hasefroch
natively (win32) using MinGW.

The Makefile is written using some GNU Make extensions (beware *BSD users,
do not forget to use "gmake").  If you are compiling within a GNU
environment (GCC), as many open source Unices are, you do not need to do
any special tricks.  Execute "make help" to find out what you need.

Commercial UNIX variants do not offer a GNU environment, mainly because
they have their proprietary C compiler, therefore compiler flags must be
modified.  However, some commercial UNIX flavours are supported as long as
GNU Make is installed.

In particular, to compile Canute in Solaris (SunOS 5.x) and in HP-UX, it is
enough to run "gmake".  Note that OSF1 (5.1 and above) testing has been
dropped because we do not have access to any such platform anymore.  Other
flavours (like AIX) may need additional tuning.  Porting patches are
welcomed.


3.1 Hasefroch (aka: Win32)
--------------------------

As a Hasefroch average user, you are not expected to be interested in this
part because we already provide a binary.  But if you would like to patch
Canute yourself and produce binaries for Hasefroch you can easily do so
from UNIX by doing cross compilation.  Install MinGW (crossed) from your
package manager and execute:

make hase

If you want to do it in Hasefroch, get GNU Make from "Unix tools for Win32"
(http://unxutils.sourceforge.net) and MinGW.  Tune up a bit the Makefile
and compile as you like.  The Makefile is very straightforward.



4. PROTOCOL ENHANCEMENTS
========================

Version 1.2 introduces some improvements on the protocol.  The previously
reserved field in the header packet now carries some additional information
about the file being transferred.

To maintain backwards compatibility with versions 1.0 and 1.1, the last
byte of each header packet is marked with a special value.  Old versions
will ignore that mark and the new information introduced into the header
packet.

So if one of the peers is using a version below 1.2, do not expect these
added features to work.


4.1. File modification time
---------------------------

Most remote copying utilities provide a way to maintain, at least, file
modification time intact.  In general it is desirable to preserve such
information for many reasons; for example, to make a backup-like remote
copy.

An attempt to use this information to decide upon resuming transfers showed
that it is trickier than it seems at a first glance.  Therefore, the resume
policy has not been altered.

Remember, though, that Canute is NOT a mirroring nor backup software.
Nevertheless, this feature helps to make Canute friendlier to such
mirroring software.


4.2. Executable bit
-------------------

The Canute experience has also shown that in many cases, the lack of
executable bit information introduces some inconveniences.  This is
particularly annoying on large directory tree transfers.  Therefore, this
information is now sent through the connection.

The executable bit is also kind of "resumed" because it propagates.  But
the lack of it does NOT propagate.  That means once the executable bit is
set locally, subsequent resumes will not clear it.

Obviously, the use of executable bit is disabled in Hasefroch builds as it
does not make sense.



5. SOURCE CODE FILES
====================

canute.h  : Dirty tricks to make the rest of the code portable and as
            #ifdef clean as possible.

canute.c  : Main function.  Command line parsing and role selection
            (server-client, sender-receiver).

feedback.c: User feedback module, progress bar, information and timing.

net.c     : Basic network management functions.  Connection handling, block
            transfer and message passing.

protocol.c: Sender-receiver negotiations and content transfers.

util.c    : Unclassified utility functions.



6. CREDITS
==========

               *** Original idea and current maintenance ***
                                  C2H5OH
                           http://www.krenel.net


                   *** Patch contributions and ideas ***
                                    MKD
                       http://www.claudiocamacho.org


                            *** Testing aid ***
                                    MKD
                                Tito Houzy
                                  m3gumi
                                   bl4d3


                        *** Initial win32 port ***
                                 Plimo San



ChangeLog
2012-02-13  etanol  <diptongo@gmail.com>

	* Canute v1.4

	- Port canute to 64-bit Windows by switching to the mingw-w64
	cross-compiler.
	- Enforce a radical transcoding of transferred file and directory
	names to printable ASCII.

2009-12-24  etanol  <diptongo@gmail.com>

	* Canute v1.3

	- Fixed a division by zero when transferring empty files (files
	with zero size).

2008-11-18  etanol  <diptongo@gmail.com>

	* Canute v1.2

	- Improved protocol, now propagating file modification time and executable
	bit information while maintaining backwards compatibility with older
	versions.
	- Makefile now detects the host platform and adjusts some
	compilation flags accordingly, so it is no longer necessary to
	perform tricks with non-gcc compilers.  Currently supported: Solaris
	and HP-UX.

2007-05-19  etanol  <diptongo@gmail.com>

	* Canute v1.1

	- Fixed a bug when giving multiple arguments on the command line.
	In some cases, the program did not return to the working directory
	correctly.
	- Progress bar now repaints at least every second, which reduces
	de number of repaints on fast connections.
	- Progress bar also can resize in the middle of a transfer.

2006-08-19  etanol  <diptongo@gmail.com>

	* Canute v1.0

	- Major version update, not backwards compatible due to
	changes in protocol.
	- Automatic resume support.
	- Large file support (> 2GBytes).
	- Some fatal errors now treated as non-fatal to gain
	robustness.
	- More UNIX flavours supported. Commercial ones tested are:
	Solaris, OSF1 (5.1) and HP-UX.
	- Split of the source file to improve code management and
	readability.
	- An informative README file added to the distribution.
	- Code licensed under GPLv2.

2006-07-16  etanol  <diptongo@gmail.com>

	* Canute v0.6

	- Modifications on time strings to make them more compact.
	- Fixed the progress bar line-jumping issue on win32.
	- Eliminated the compiler warning in function gettimeofday()
	on win32.

2006-07-02  etanol  <diptongo@gmail.com>

	* Canute v0.5

	- More detailed feedback before and after progress bar. Thanks
	to a small code redesign.
	- Stabilized speed and ETA values after some recoded parts in
	order to find and resolve a stability bug found when
	transmitting over slow connections.
	- Cleaned some unneeded conditional code (some #ifdef's supressed).

2006-06-27  mkd  <claudiomkd@gmail.com>

	* Canute v0.4.1

	- Add a small workaround for ETA calculations when the file
	size is less than two megabytes.


2006-06-26  mkd  <claudiomkd@gmail.com>

	* Canute v0.4

	- New feature: A mean speed is shown after each transfer.
	- New timer code for Windows using an equivalent to the Unix.
	gettimeofday().
	- Include some stability changes to the global timer code.
	- Some code clean-ups and remove non-necessary variables.

2006-05-19  etanol  <diptongo@gmail.com>

	* Canute v0.3

	- Fixed a small bug in the speed meter which made the whole
	progress bar to jump down a line and produced an ugly visual
	effect.

2006-05-15  etanol  <diptongo@gmail.com>

	* Canute v0.2

	- Improved feedback. Now we have a WGet style progress bar,
	most of the code was stolen from there anyway. The transmission
	speed and estimated time needs some improvements, basically
	stabilize the values.

2005-12-03  etanol  <diptongo@gmail.com>

	* Canute v0.1

	- First stable release with rudimentary feedback.



net.c
/******************************************************************************/
/*                ____      _      _   _   _   _   _____   _____              */
/*               / ___|    / \    | \ | | | | | | |_   _| | ____|             */
/*              | |       / _ \   |  \| | | | | |   | |   |  _|               */
/*              | |___   / ___ \  | |\  | | |_| |   | |   | |___              */
/*               \____| /_/   \_\ |_| \_|  \___/    |_|   |_____|             */
/*                                                                            */
/*               NETWORK MANAGEMENT AND TRANSMISSION FUNCTIONS                */
/*                                                                            */
/******************************************************************************/

#include "canute.h"


/*
 * open_connection_server
 *
 * Set up a connection in server mode. Opens the specified port for listening
 * and wait for someone to connect. Return the connected socket ready for
 * transmission.
 */
SOCKET open_connection_server (unsigned short port)
{
        SOCKET             bsk, sk; /* bsk --> Binding SocKet */
        struct sockaddr_in saddr;
        int                e;
        socklen_t          alen;

        bsk = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
        if (bsk == INVALID_SOCKET)
                fatal("Could not create socket");

        saddr.sin_family      = AF_INET;
        saddr.sin_port        = htons(port);
        saddr.sin_addr.s_addr = INADDR_ANY;

        /* Ignore errors from setsockopt(), bind() will fail in that case */
        e = 1;
        setsockopt(bsk, SOL_SOCKET, SO_REUSEADDR, CCP_CAST &e, sizeof(e));

        e = bind(bsk, (SOCKADDR *) &saddr, sizeof(saddr));
        if (e == SOCKET_ERROR)
                fatal("Could not open port %d", port);

        e = listen(bsk, 1);
        if (e == SOCKET_ERROR)
                fatal("Could not listen on port %d", port);

        alen = sizeof(saddr);
        sk   = accept(bsk, (SOCKADDR *) &saddr, &alen);
        if (sk == INVALID_SOCKET)
                fatal("Could not accept client connection");

        /* Binding socket not needed anymore */
        closesocket(bsk);
        return sk;
}


/*
 * open_connection_client
 *
 * Set up a connection in client mode. Try to connect to the specified host
 * (either a hostname or an IP address) and return the connected socket ready
 * for transmission.
 */
SOCKET open_connection_client (char *host, unsigned short port)
{
        SOCKET             sk;
        struct sockaddr_in saddr;
        struct hostent    *he;
        int                e;

        sk = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
        if (sk == INVALID_SOCKET)
                fatal("Creating socket");

        saddr.sin_family      = AF_INET;
        saddr.sin_port        = htons(port);
        saddr.sin_addr.s_addr = inet_addr(host);
        if (saddr.sin_addr.s_addr == INADDR_NONE)
        {
                /* Invalid IP, maybe we have to do a DNS query */
                he = gethostbyname(host);
                if (he == NULL)
                {
#if defined(HASEFROCH) || defined(OMIT_HERROR)
                        fatal("Invalid IP or hostname");
#else
                        herror("FATAL ERROR: Invalid IP or hostname");
                        exit(EXIT_FAILURE);
#endif
                }
                saddr.sin_addr.s_addr = ((struct in_addr *) he->h_addr)->s_addr;
        }

        /* Now we have a destination host */
        e = connect(sk, (SOCKADDR *) &saddr, sizeof(saddr));
        if (e == SOCKET_ERROR)
                fatal("Connecting to host '%s'", host);

        return sk;
}


/*
 * send_data
 *
 * Send count bytes over the connection. Doesn't finish until all of them are
 * sent. On error aborts.
 */
void send_data (SOCKET sk, char *buf, size_t count)
{
        int s; /* Sent bytes in one send() call */

        do {
                s = send(sk, buf, count, 0);
                if (s == SOCKET_ERROR)
                        fatal("Sending data");
                count -= s;
                buf   += s;
        } while (count > 0);
}


/*
 * receive_data
 *
 * Receive count bytes from the connection. Doesn't finish until all bytes are
 * received. On error aborts.
 */
void receive_data (SOCKET sk, char *buf, size_t count)
{
        int r; /* Received bytes in one recv() call */

        do {
                r = recv(sk, buf, count, 0);
                if (r == SOCKET_ERROR)
                        fatal("Receiving data");
                count -= r;
                buf   += r;
        } while (count > 0);
}


/*
 * send_message
 *
 * Build a header packet and send it through the connection. All the fields are
 * converted to network byte order if required.
 */
void send_message (SOCKET    sk,
                   int       type,
                   int       is_executable,
                   int       mtime,
                   long long size,
                   char     *name)
{
        int           blocks, extra;
        struct header packet;

        blocks = (int) (size >> CANUTE_BLOCK_BITS);
        extra  = (int) (size &  CANUTE_BLOCK_MASK);

        /* Encode executable bit on the mtime filed;  never mind compatibility
         * because the remote side will ignore the mtime field */
        if (is_executable)
                mtime = -mtime;

        packet.type   = htonl(type);
        packet.mtime  = htonl(mtime);
        packet.blocks = htonl(blocks);  /* Read protocol.c for an explanation */
        packet.extra  = htonl(extra);

        if (name != NULL)
                strncpy(packet.name, name, CANUTE_NAME_LENGTH);
        else
                packet.name[0] = '\0';

        /* Mark the packet as enhanced version and send it */
        packet.name[CANUTE_NAME_LENGTH] = CANUTE_ENHANCED;
        send_data(sk, (char *) &packet, sizeof(struct header));
}


/*
 * receive_message
 *
 * Read from the connection expecting a header packet. Fix byte ordering if
 * necessary, fill the fields (if address was provided by the caller) and return
 * the message type.
 */
int receive_message (SOCKET     sk,
                     int       *is_executable,
                     int       *mtime,
                     long long *size,
                     char      *name)
{
        int           blocks, extra, pmtime = 0, is_x = 0;
        struct header packet;

        receive_data(sk, (char *) &packet, sizeof(struct header));

        if (packet.name[CANUTE_NAME_LENGTH] == CANUTE_ENHANCED)
        {
                /* Decode executable bit from the mtime field */
                pmtime = ntohl(packet.mtime);
                if (pmtime < 0)
                {
                        is_x   = !is_x;
                        pmtime = -pmtime;
                }
        }

        if (is_executable != NULL)
                *is_executable = is_x;

        if (mtime != NULL)
                *mtime = pmtime;

        if (size != NULL)
        {
                /* Read protocol.c for an explanation */
                blocks = ntohl(packet.blocks);
                extra  = ntohl(packet.extra);
                *size  = ((long long) blocks << 16) + extra;
        }

        if (name != NULL)
        {
                strncpy(name, packet.name, CANUTE_NAME_LENGTH);
                name[CANUTE_NAME_LENGTH] = '\0';
        }

        return ntohl(packet.type);
}



file.c
/******************************************************************************/
/*                ____      _      _   _   _   _   _____   _____              */
/*               / ___|    / \    | \ | | | | | | |_   _| | ____|             */
/*              | |       / _ \   |  \| | | | | |   | |   |  _|               */
/*              | |___   / ___ \  | |\  | | |_| |   | |   | |___              */
/*               \____| /_/   \_\ |_| \_|  \___/    |_|   |_____|             */
/*                                                                            */
/*                            FILESYSTEM FUNCTIONS                            */
/*                                                                            */
/******************************************************************************/

#include "canute.h"

#ifdef HASEFROCH
/*
 * Seems that this prototype is missing in MinGW headers. Even though the symbol
 * exists and is linkable.
 */
extern int _stat64 (const char *path, struct __stat64 *buffer);
#endif


/*
 * basename
 *
 * Strip absolute path. Similar to UNIX "basename" command but with a somewhat
 * dirty implementation.
 */
char *
basename (char *path)
{
#ifdef HASEFROCH
#define PATH_SEPARATOR '\\'
#else
#define PATH_SEPARATOR '/'
#endif
        int p = strlen(path) - 1;

        while (path[p] == PATH_SEPARATOR) {
                path[p] = '\0';
                p--;
        }
        while (p > 0 && path[p] != PATH_SEPARATOR)
                p--;
        if (path[p] == PATH_SEPARATOR)
                p++;
        return path + p;
}


/*
 * file_size
 *
 * Return path's size in bytes. Report a zero size in case of an error (file
 * unavailable, permission denied, etc).
 *
 * It comes handy to return 0 as an error because then empty files (with zero
 * size) are treated as non-existent, then skipped by the sender. From the
 * pragmatic point of view is desirable because files without content don't have
 * any interest.
 *
 * From the receiver perspective it is also handy because if the file doesn't
 * exist then the whole file is requested (zero initial offset).
 */
int64_t
file_size (char *path)
{
        int e;
#ifdef HASEFROCH
        struct __stat64 st;
        e = _stat64(path, &st);
#else
        struct stat st;
        e = stat(path, &st);
#endif

        return (e == -1 ? 0LL : (int64_t) st.st_size);
}


/*
 * change_wd
 *
 * Change working directory, similar to "cd" command. Abort on error. Specially
 * important if used just after a try_mkdir() because its silent errors will be
 * noticeable here.
 */
void
change_wd (char *path)
{
        int e;

        e = chdir(path);
        if (e == -1)
                fatal("Changing current directory to '%s'", path);
}


/*
 * try_mkdir
 *
 * Create a directory. But fail silently if it couldn't be created.
 */
void
try_mkdir (char *path)
{
#ifdef HASEFROCH
        mkdir(path);
#else
        mkdir(path, 0755);
#endif
}


#ifdef HASEFROCH
/*
 * fseeko
 *
 * As fsetpos() is also implemented in UNIX, that function could have been used
 * instead. But fpos_t is a composite type (struct) rather than a 64 bit integer
 * as in Hasefroch.
 */
int
fseeko (FILE *stream, off_t offset, int whence)
{
        int    e;
        fpos_t pos;

        if (whence != SEEK_SET)
                return -1;

        pos = (fpos_t) offset;
        e   = fsetpos(stream, &pos);
        if (e != 0)
                return -1;

        return 0;
}
#endif



protocol.c
/******************************************************************************/
/*                ____      _      _   _   _   _   _____   _____              */
/*               / ___|    / \    | \ | | | | | | |_   _| | ____|             */
/*              | |       / _ \   |  \| | | | | |   | |   |  _|               */
/*              | |___   / ___ \  | |\  | | |_| |   | |   | |___              */
/*               \____| /_/   \_\ |_| \_|  \___/    |_|   |_____|             */
/*                                                                            */
/*                          PROTOCOL IMPLEMENTATION                           */
/*                                                                            */
/******************************************************************************/

/*
 * EXPLANATION
 *
 * The protocol used (if could be labeled as "protocol") is very simple. There
 * is a header packet (struct header) which contains the packet or message type,
 * the size in a custom representation and the name of the item. A header is
 * always sent to make the receiver host know the size of the contents when
 * required. Even though the same data structure is used for all messages, the
 * sender's messages are requests and the receiver's are replies.
 *
 * There are four types of requests (sender) and two types of replies
 * (receiver); the requests identify the kind of item that has to be negotiated.
 * In some cases the receiver must answer the sender requests. Those messages
 * are identified by a self-explanatory constant.
 *
 * When a file is about to be transferred, the sender uses a REQUEST_FILE. The
 * receiver must reply this message with a REPLY_ACCEPT if it wants the file,
 * indicating in the same message the initial offset of the file (resume
 * support). In case the receiver doesn't want or need the file, it must answer
 * with a REPLY_SKIP.
 *
 * If the sender needs to send a directory, a REQUEST_BEGINDIR with the name of
 * the directory has to be sent. The receiver must accept or skip the directory
 * (directory skipping normally happens on error situations). When a directory
 * has been accepted is walked through recursively sending each directory entry
 * with the corresponding header. In order to make directory recursion easier,
 * after a directory has been accepted, both peers should move into it.
 *
 * After the directory has been completely read, a REQUEST_ENDDIR must be sent
 * and both peers should return to the parent directory. This "end of directory"
 * request must NOT be answered and any error occured when moving to parent
 * directory is considered fatal.
 *
 * Finally, when no more items are left, a REQUEST_END is sent to notify the
 * receiver about the situation. Then both hosts close the connection and
 * exit.
 *
 *
 * ABOUT FILE SIZES
 *
 * When large file support (LFS) came into scene, some issues arised. The most
 * important was the handling of 64 bit numbers in 32 bit machines. Internal
 * details about representation and handling (parameter passing, etc) were not
 * clear and could obviously vary from platform to platform.
 *
 * Initial tests failed because middle endianness in 64 bit numbers was
 * discarded. And there isn't too much documentation about it. Having a beach
 * bath with MKD (see README), he gave me the idea of using a multiplier as the
 * transfer was block based. With this approach, up to 47 bits (without sign)
 * can be used to represent a size, which can represent 128 TeraBytes of
 * storage.
 *
 * This way we can transfer large sizes without worrying about local byte order
 * (ntohl and htonl are transparent) and 64 bit number representation.
 */
#include "canute.h"

static char databuf[CANUTE_BLOCK_SIZE];


/****************************  PRIVATE FUNCTIONS  ****************************/

/*
 * receive_file
 *
 * A file request has been received from the network.  We must reply depending
 * on the local state of the file requested.
 *
 * Having mtime > 0 means that the peer is version above 1.1, so we can set the
 * file mtime to that provided by the protocol.
 */
static void receive_file (SOCKET    sk,
                          char     *name,
                          long long size,
                          int       mtime,
                          int       is_executable)
{
        int               e;
        FILE             *file;
        long long         received_bytes; /* Think about it also as "offset" */
        size_t            b;
        struct stat_info  st;
        struct utime_info ut;

        e = stat(name, &st);
        if (e == -1)
                received_bytes = 0;  /* Most probable: errno == ENOENT */
        else if (st.st_size >= size)
        {
                printf("--- Skipping file '%s'\n", name);
                send_message(sk, REPLY_SKIP, 0, 0, 0, NULL);
                return;
        }
        else
                received_bytes = (long long) st.st_size;

        file = fopen(name, (received_bytes > 0 ? "ab" : "wb"));
        if (file == NULL)
        {
                error("Cannot open file '%s'", name);
                send_message(sk, REPLY_SKIP, 0, 0, 0, NULL);
                return;
        }

        send_message(sk, REPLY_ACCEPT, 0, 0, received_bytes, NULL);
        setup_progress(name, size, received_bytes);

        while (received_bytes < size)
        {
                if (size - received_bytes > CANUTE_BLOCK_SIZE)
                        b = CANUTE_BLOCK_SIZE;
                else
                        b = (size_t) (size - received_bytes);

                receive_data(sk, databuf, b);
                fwrite(databuf, 1, b, file);
                update_progress(b);
                received_bytes += b;
        }

        finish_progress();
        fflush(file);
        fclose(file);

        /* Set mtime if packet provides information */
        if (mtime > 0)
        {
                ut.actime  = (time_t) mtime;
                ut.modtime = (time_t) mtime;
                e = utime(name, &ut);
                if (e == -1)
                        error("Cannot set modification time on '%s'", name);
        }

#ifndef HASEFROCH
        if (is_executable)
        {
                e = stat(name, &st);
                if (e != -1)
                {
                        e = chmod(name, st.st_mode | S_IXUSR);
                        if (e == -1)
                                error("Setting executable bit on '%s'", name);
                }
                else
                        error("Cannot stat file '%s'", name);
        }
#endif
}


/*
 * send_file
 *
 * Treat the item as a file and try to send it.
 */
static void send_file (SOCKET    sk,
                       char     *name,
                       long long size,
                       int       mtime,
                       int       is_executable)
{
        int       e, reply;
        long long sent_bytes; /* Size reported remotely */
        size_t    b;
        char     *sname;
        FILE     *file;

        file  = fopen(name, "rb");
        if (file == NULL)
        {
                error("Cannot open file '%s'", name);
                return;
        }

        sname = safename(name);
        send_message(sk, REQUEST_FILE, is_executable, mtime, size, sname);
        reply = receive_message(sk, NULL, NULL, &sent_bytes, NULL);
        if (reply == REPLY_SKIP)
        {
                fclose(file);
                printf("--- Skipping file '%s'\n", sname);
                return;
        }

        if (sent_bytes > 0)
        {
                e = fseeko(file, (off_t) sent_bytes, SEEK_SET);
                if (e == -1)
                        fatal("Could not seek file '%s'", sname);
        }

        setup_progress(sname, size, sent_bytes);

        while (sent_bytes < size)
        {
                b = fread(databuf, 1, CANUTE_BLOCK_SIZE, file);
                send_data(sk, databuf, b);
                update_progress(b);
                sent_bytes += b;
        }

        finish_progress();
        fclose(file);
}


/*****************************  PUBLIC FUNCTIONS  *****************************/

/*
 * send_item
 *
 * Discover what kind of filesystem item 'name' represents and send it over the
 * connection.
 */
void send_item (SOCKET sk, char *name)
{
        int              e, reply, x_bit = 0;
        char            *sname;
        DIR             *dir;
        struct dirent   *dentry;
        struct stat_info st;

        e = stat(name, &st);
        if (e == -1)
        {
                error("Cannot stat item '%s'", name);
                return;
        }

        if (S_ISDIR(st.st_mode))
        {
                dir = opendir(name);
                if (dir == NULL)
                {
                        error("Cannot open dir '%s'", name);
                        return;
                }

                e = chdir(name);
                if (e == -1)
                {
                        closedir(dir);
                        error("Cannot change to dir '%s'", name);
                        return;
                }

                sname = safename(name);
                send_message(sk, REQUEST_BEGINDIR, 0, 0, 0, sname);
                reply = receive_message(sk, NULL, NULL, NULL, NULL);
                if (reply == REPLY_SKIP)
                {
                        closedir(dir);
                        printf("--- Skipping directory '%s'\n", sname);
                        e = chdir("..");
                        if (e == -1)
                                fatal("Could not change to parent directory");
                        return;
                }

                printf(">>> Entering directory '%s'\n", sname);
                dentry = readdir(dir);
                while (dentry != NULL)
                {
                        if (NOT_SELF_OR_PARENT(dentry->d_name))
                                send_item(sk, dentry->d_name);
                        dentry = readdir(dir);
                }

                closedir(dir);
                e = chdir("..");
                if (e == -1)
                        fatal("Could not change to parent directory");
                send_message(sk, REQUEST_ENDDIR, 0, 0, 0, NULL);
        }
        else
        {
#ifndef HASEFROCH
                x_bit = st.st_mode & S_IXUSR;
#endif
                send_file(sk, name, st.st_size, (int) st.st_mtime, x_bit);
        }
}


/*
 * receive_item
 *
 * Parse a header packet containing a request and act accordingly. Return true
 * if the read packet is the last one for the session (no more items to come)
 * and false otherwise (still more stuff pending).
 */
int receive_item (SOCKET sk)
{
        static char namebuf[CANUTE_NAME_LENGTH];
        int         e, x_bit, mtime, request;
        long long   size;

        request = receive_message(sk, &x_bit, &mtime, &size, namebuf);

        switch (request)
        {
        case REQUEST_FILE:
                receive_file(sk, namebuf, size, mtime, x_bit);
                break;

        case REQUEST_BEGINDIR:
                mkdir(namebuf);
                e = chdir(namebuf);
                if (e == -1)
                {
                        error("Cannot change to dir '%s'", namebuf);
                        send_message(sk, REPLY_SKIP, 0, 0, 0, NULL);
                }
                else
                {
                        printf(">>> Entering directory '%s'\n",  namebuf);
                        send_message(sk, REPLY_ACCEPT, 0, 0, 0, NULL);
                }
                break;

        case REQUEST_ENDDIR:
                e = chdir("..");
                if (e == -1)
                        fatal("Could not change to parent directory");
                break;

        case REQUEST_END:
                return 1;

        default:
                fatal("Unexpected header type (%d)", request);
        }

        return 0;
}

